### PR TITLE
docs: add architecture doc

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have added myself to the contributors in `CHANGELOG.md` (optional)

### CODE REVIEW

**Review by chooglen — APPROVED** [review]
LGTM. Some things could be clearer but we can clean that up later.

File: docs/technical/architecture.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+## Design of the library crate
+
+### Overview
+
+Here's a diagram showing some important types in the library crate. The
+following sections describe each component.
+
+```mermaid
+graph TD;
+    ReadonlyRepo-->Store;
```
[the remark this answers is not in the document]
Comment (martinvonz) [reply docs/technical/architecture.md:54]: Makes sense

File: docs/technical/architecture.md [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    RepoLoader-->OpHeadsStore;
+    RepoLoader-->ReadonlyRepo;
+    Git-->GitBackend;
+    GitBackend-->StackedTable;
+```
+
+### StackedTable
+
+`StackedTable` (actually `ReadonlyTable` and `MutableTable`) is a simple disk
+format for storing key-value pairs sorted by key. The keys have to have the same
```
[the remark this answers is not in the document]
Comment (martinvonz) [reply docs/technical/architecture.md]: I moved it to the end for now.

File: docs/technical/architecture.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+size but the values can have different sizes. We use a custom format because we
+want [lock-free concurrency](concurrency.md).
```
[the remark this answers is not in the document]
Comment (martinvonz) [reply docs/technical/architecture.md]: As opposed to any existing format :) Updated

**Review by PhilipMetzger — APPROVED** [review]
LG, maybe we could mention the other upcoming index types (Conflict- and Tree)? As discussed in this [doc](https://docs.google.com/document/d/1NVfcFQvl9CnBoL63hoC3iu85nWucDtHNfhSRof88iF8/edit#).

**Pull request merged**